Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -23,586 +23,642 @@
+
+    @assert(stack.@filter(e => e === entry).length === 1);
+    @assert(entry.dfsAncestorIndex <= entry.dfsIndex);
+
+    if (entry.dfsAncestorIndex === entry.dfsIndex) {
+        for (;;) {
+            var requiredModule = stack.@pop();
+            requiredModule.status = @ModuleStatusLinked;
+            if (requiredModule === entry)
+                break;
+        }
+    }
+
+    return index;
 }
 
+// https://tc39.es/ecma262/#sec-moduleevaluation
 @visibility=PrivateRecursive
-function requestSatisfyUtil(entry, parameters, fetcher, visited, satisfyingEntries)
+function moduleEvaluation(entry, fetcher)
 {
-    // https://html.spec.whatwg.org/#internal-module-script-graph-fetching-procedure
+    if (entry.errorToRethrow) {
+        var capability = @newPromiseCapability(@InternalPromise);
+        capability.reject(entry.errorToRethrow);
+        return capability.promise;
+    }
 
-    "use strict";
+    @assert(entry.status === @ModuleStatusLinked || entry.status === @ModuleStatusEvaluatingAsync || entry.status === @ModuleStatusEvaluated);
 
-    if (entry.satisfy)
-        return entry.satisfy;
-
-    visited.@add(entry);
-    var satisfyPromise = this.requestInstantiate(entry, parameters, fetcher).then((entry) => {
-        if (entry.satisfy)
-            return entry.satisfy;
-
-        var depLoads = this.requestedModuleParameters(entry.module);
-        for (var i = 0, length = entry.dependencies.length; i < length; ++i) {
-            var parameters = depLoads[i];
-            var depEntry = entry.dependencies[i];
-            var promise;
-
-            // Recursive resolving. The dependencies of this entry is being resolved or already resolved.
-            // Stop tracing the circular dependencies.
-            // But to retrieve the instantiated module record correctly,
-            // we need to wait for the instantiation for the dependent module.
-            // For example, reaching here, the module is starting resolving the dependencies.
-            // But the module may or may not reach the instantiation phase in the loader's pipeline.
-            // If we wait for the Satisfy for this module, it construct the circular promise chain and
-            // rejected by the Promises runtime. Since only we need is the instantiated module, instead of waiting
-            // the Satisfy for this module, we just wait Instantiate for this.
-            if (visited.@has(depEntry))
-                promise = this.requestInstantiate(depEntry, parameters, fetcher);
-            else {
-                // Currently, module loader do not pass any information for non-top-level module fetching.
-                promise = this.requestSatisfyUtil(depEntry, parameters, fetcher, visited, satisfyingEntries);
-            }
-            @putByValDirect(depLoads, i, promise);
+    // https://searchfox.org/mozilla-central/rev/1f5e1875cbfd5d4b1bfa27ca54832f62dd19589e/js/src/vm/Modules.cpp#1431
```

Comment:
Maybe we should state the reason here and not provide the link to other's codebase?